File: src/main/java/com/zunza/buythedip/auth/oauth2/CustomOAuth2User.java
package com.zunza.buythedip.auth.oauth2;

import java.util.Collection;
import java.util.List;
import java.util.Map;

import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.oauth2.core.user.OAuth2User;

import com.zunza.buythedip.user.entity.User;

import lombok.RequiredArgsConstructor;

@RequiredArgsConstructor
public class CustomOAuth2User implements OAuth2User {
	private final User user;

	@Override
	public Map<String, Object> getAttributes() {
		return Map.of();
	}

	@Override
	public Collection<? extends GrantedAuthority> getAuthorities() {
		return List.of((GrantedAuthority)() -> user.getUserRole().getValue());
	}

	@Override
	public String getName() {
		return user.getNickname();
	}

	public Long getUserId() {
		return user.getId();
	}
}

File: src/main/kotlin/com/zunza/buythedip/auth/oauth2/CustomOAuth2User.kt
package com.zunza.buythedip.auth.oauth2

import com.zunza.buythedip.user.entity.User
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.oauth2.core.user.OAuth2User

class CustomOAuth2User(
    private val user: User
) : OAuth2User {

    companion object {
        fun createOf(user: User): CustomOAuth2User {
            return CustomOAuth2User(user)
        }
    }

    override fun getAttributes(): Map<String?, Any?>? {
        return mapOf()
    }

    override fun getAuthorities(): Collection<GrantedAuthority?>? {
        return listOf(GrantedAuthority { user.role.value })
    }

    override fun getName(): String? {
        return user.nickname
    }

    val email: String
        get() = user.email

    val id: Long
        get() = user.id
}
